### PR TITLE
feat: implement detection endpoint

### PR DESCRIPTION
**Pull request by Dinasoa**

### CODE REVIEW

File: doc/api.yml [hunk 1]
```diff
@@ -996,3 +1016,31 @@ components:
...
+      allOf:
+        - $ref: '#/components/schemas/CreateZoneTilingJob'
+      properties:
+        objectType:
+          $ref: '#/components/schemas/DetectableObjectType'
+
+    Type:
+      enum:
+        - MACHINE
+        - HUMAN
```
Comment (hei-teacher) [inline doc/api.yml]: HUMAN_DETECTION

File: doc/api.yml [hunk 2]
```diff
@@ -996,3 +1016,31 @@ components:
...
+      type: object
+      allOf:
+        - $ref: '#/components/schemas/CreateZoneTilingJob'
+      properties:
+        objectType:
+          $ref: '#/components/schemas/DetectableObjectType'
+
+    Type:
+      enum:
+        - MACHINE
```
Comment (hei-teacher) [inline doc/api.yml]: MACHINE_DETECTION

File: doc/api.yml [hunk 3]
```diff
@@ -996,3 +1016,31 @@ components:
...
+      properties:
+        type:
+          type: array
+          items:
+           $ref: '#/components/schemas/Type'
+        statistics:
+          type: array
+          items:
+           $ref: '#/components/schemas/TaskStatistic'
+        geojsonUrl:
```
Comment (hei-teacher) [inline doc/api.yml]: detectedGeojsonUrl

File: doc/api.yml [hunk 4]
```diff
@@ -996,3 +1016,31 @@ components:
...
+        - HUMAN
+        - TILING
+
+    DetectedZone:
+      type: object
+      properties:
+        type:
+          type: array
+          items:
+           $ref: '#/components/schemas/Type'
```
Comment (hei-teacher) [inline doc/api.yml]: Type of what !!! Jesus, can you stop being so NOT explicit!!!

File: doc/api.yml [hunk 5]
```diff
@@ -464,6 +464,26 @@ paths:
                 $ref: "#/components/schemas/ZoneDetectionJob"
         '500':
           $ref: '#/components/responses/500'
+
+  '/detection':
```
Comment (hei-teacher) [inline doc/api.yml]: FullDetection. And use that name everywhere

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 6]
```diff
@@ -4,6 +4,7 @@ build/
 !gradle/wrapper/gradle-wrapper.jar
 !**/src/main/**/build/
 !**/src/test/**/build/
+venv/
```
Comment (hei-ryan) [inline .gitignore]: Why ?

File: .gitignore [hunk 7]
```diff
@@ -40,4 +41,4 @@ out/
 **.toml
 
 ### Mac OS ###
-.DS_Store
\ No newline at end of file
+.DS_Store
```
Comment (hei-ryan) [inline .gitignore]: No

**Review by hei-ryan — CHANGES_REQUESTED** [review]

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java [hunk 8]
```diff
@@ -22,6 +23,12 @@ public class TilingTaskCreated extends PojaEvent {
   @JsonProperty("tilingTask")
   private TilingTask task;
 
+  private CreateFullDetection fullDetection;
+
+  public TilingTaskCreated(TilingTask task) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java:28]: Is this really required?I don't think so

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java [hunk 9]
```diff
@@ -22,6 +23,12 @@ public class TilingTaskCreated extends PojaEvent {
   @JsonProperty("tilingTask")
   private TilingTask task;
 
+  private CreateFullDetection fullDetection;
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java:26]: Do not forget to add `@JsonProperty("")` for consistency

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskSucceeded.java [hunk 10]
```diff
@@ -19,6 +20,11 @@
 @ToString
 public class TilingTaskSucceeded extends PojaEvent {
   private TilingTask task;
+  private CreateFullDetection fullDetection;
+
+  private TilingTaskSucceeded(TilingTask tilingTask) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskSucceeded.java]: same

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskFailed.java [hunk 11]
```diff
@@ -20,6 +21,7 @@
 public class TilingTaskFailed extends PojaEvent {
   private TilingTask task;
   private int attemptNb;
+  private CreateFullDetection createFullDetection;
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskFailed.java:24]: same

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/ZTJStatusRecomputingSubmitted.java [hunk 12]
```diff
@@ -11,6 +12,15 @@
 public class ZTJStatusRecomputingSubmitted extends JobStatusRecomputingSubmitted {
   private static final long MAX_CONSUMER_DURATION_VALUE = 300L;
   private static final long DEFAULT_BACK_OFF_VALUE = 180L;
+  private CreateFullDetection createFullDetection;
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/ZTJStatusRecomputingSubmitted.java]: make it `final`

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/ZoneTilingJobCreated.java [hunk 13]
```diff
@@ -22,6 +23,12 @@ public class ZoneTilingJobCreated extends PojaEvent {
   @JsonProperty("zoneTilingJob")
   private ZoneTilingJob zoneTilingJob;
 
+  private CreateFullDetection createFullDetection;
+
+  public ZoneTilingJobCreated(ZoneTilingJob zoneTilingJob) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/ZoneTilingJobCreated.java:28]: why?

File: src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java [hunk 14]
```diff
@@ -12,15 +13,19 @@
...
 
   public DetectableObjectConfiguration toDomain(
       String jobId, app.bpartners.geojobs.endpoint.rest.model.DetectableObjectConfiguration rest) {
     return DetectableObjectConfiguration.builder()
         .id(randomUUID().toString())
         .detectionJobId(jobId)
         .objectType(typeMapper.toDomain(Objects.requireNonNull(rest.getType())))
-        .confidence(Objects.requireNonNull(rest.getConfidence()).doubleValue())
-        .bucketStorageName(rest.getBucketStorageName())
+        .confidence(rest.getConfidence() != null ? rest.getConfidence().doubleValue() : 1)
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java:25]: what is this magic number `1`?

File: src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java [hunk 15]
```diff
@@ -29,8 +34,8 @@ public app.bpartners.geojobs.endpoint.rest.model.DetectableObjectConfiguration t
     return new app.bpartners.geojobs.endpoint.rest.model.DetectableObjectConfiguration()
         .confidence(
             domain.getConfidence() == null
-                ? null
-                : BigDecimal.valueOf(domain.getConfidence())) // TODO: Unknown default value
+                ? BigDecimal.valueOf(1)
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java]: put `1` in a static variable `DEFAULT_CONFIDENCE` for example

File: src/main/java/app/bpartners/geojobs/service/ZoneService.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+        zoneDetectionJobService.processZDJ(
+            zoneDetectionJob.getId(), detectableObjectConfigurations);
+
+    JobStatus jobStatus = processedZDJ.getStatus();
+    if (!FINISHED.equals(jobStatus.getProgression()) && !SUCCEEDED.equals(jobStatus.getHealth())) {
+      eventProducer.accept(List.of(new ZDJStatusRecomputingSubmitted(processedZDJ.getId())));
+    }
+  }
+
+  public boolean isPending(ZoneDetectionJob job) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/service/ZoneService.java]: No need. I think this method already exists for a job: `job.isPending()`

File: src/main/java/app/bpartners/geojobs/service/event/TileDetectionTaskSucceededService.java [hunk 17]
```diff
@@ -12,11 +19,27 @@
...
     tileDetectionTaskStatusService.succeed(tileDetectionTask);
+    List<TileDetectionTask> tasks =
+        tileDetectionTaskRepository.findAllByJobId(tileDetectionTask.getJobId());
+    if (isFinished(tasks)) {
+      eventProducer.accept(
+          List.of(new ZDJStatusRecomputingSubmitted(tileDetectionTask.getJobId())));
+    }
+  }
+
+  private boolean isFinished(List<TileDetectionTask> tasks) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/service/event/TileDetectionTaskSucceededService.java]: Create it in the `Task` class if not exists

File: src/main/java/app/bpartners/geojobs/service/event/ZDJStatusRecomputingSubmittedService.java [hunk 18]
```diff
@@ -32,5 +43,10 @@ public ZDJStatusRecomputingSubmittedService(
   @Override
   public void accept(ZDJStatusRecomputingSubmitted event) {
     service.accept(event);
+    ZoneDetectionJob zoneDetectionJob = zoneDetectionJobService.findById(event.getJobId());
+    JobStatus jobStatus = zoneDetectionJob.getStatus();
+    if (FINISHED.equals(jobStatus.getProgression()) && SUCCEEDED.equals(jobStatus.getHealth())) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/service/event/ZDJStatusRecomputingSubmittedService.java]: replace with `zoneDetectionJob.isSucceeded()`

File: src/main/java/app/bpartners/geojobs/service/event/ZTJStatusRecomputingSubmittedService.java [hunk 19]
```diff
@@ -30,5 +45,15 @@ public ZTJStatusRecomputingSubmittedService(
...
   public void accept(ZTJStatusRecomputingSubmitted event) {
     service.accept(event);
+    Optional<ZoneTilingJob> zoneTilingJob = zoneTilingJobRepository.findById(event.getJobId());
+    if (zoneTilingJob.isEmpty()) {
+      throw new ApiException(
+          SERVER_EXCEPTION, String.format("The job %s is not found", event.getJobId()));
+    }
+    ZoneTilingJob ZTJ = zoneTilingJob.get();
+    JobStatus jobStatus = ZTJ.getStatus();
+    if (FINISHED.equals(jobStatus.getProgression()) && SUCCEEDED.equals(jobStatus.getHealth())) {
```
Comment (nvagno) [inline src/main/java/app/bpartners/geojobs/service/event/ZTJStatusRecomputingSubmittedService.java]: same

**Review by nvagno — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 20]
```diff
@@ -40,4 +40,5 @@ out/
 **.toml
 
 ### Mac OS ###
-.DS_Store
```
Comment (hei-ryan) [inline .gitignore]: Don't touch

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/ZTJStatusRecomputingSubmitted.java (hunk 12 above)
Comment (hei-ryan) [reply src/main/java/app/bpartners/geojobs/endpoint/event/model/ZTJStatusRecomputingSubmitted.java]: Why did you resolve without making it final ?

File: src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java [hunk 21]
```diff
@@ -12,15 +13,20 @@
 @AllArgsConstructor
 public class DetectableObjectConfigurationMapper {
   private final DetectableObjectTypeMapper typeMapper;
+  public static String BUCKET_STORAGE_NAME = "TILING_IMAGES";
```
Comment (hei-ryan) [inline src/main/java/app/bpartners/geojobs/endpoint/rest/controller/mapper/DetectableObjectConfigurationMapper.java]: Why is it a constant inside the code here ?

File: src/main/resources/db/migration/V43_50__Create_full_detection_table.sql [hunk 22]
```diff
@@ -5,4 +5,4 @@ create table full_detection (
     zdj_id varchar references zone_detection_job("id"),
     detectable_object_configuration jsonb,
     geojson_s3_file_key varchar
-);
+);
```
Comment (hei-ryan) [inline src/main/resources/db/migration/V43_50__Create_full_detection_table.sql]: Undo it

File: src/main/java/app/bpartners/geojobs/endpoint/rest/controller/ZoneDetectionController.java [hunk 23]
```diff
@@ -141,17 +145,17 @@ public app.bpartners.geojobs.endpoint.rest.model.ZoneDetectionJob processZDJ(
...
     return mapper.toRest(processedZDJ, detectableObjectConfigurations);
   }
 
   @GetMapping("/detectionJobs/{id}/geojsonsUrl")
   public GeoJsonsUrl getZDJGeojsonsUrl(@PathVariable(value = "id") String detectionJobId) {
     return geoJsonConversionInitiationService.initiateGeoJsonConversion(detectionJobId);
   }
+
+  @PutMapping("/fullDetection")
+  public FullDetectedZone processFullDetection(@RequestBody CreateFullDetection zoneToDetect) {
```
Comment (hei-ryan) [inline src/main/java/app/bpartners/geojobs/endpoint/rest/controller/ZoneDetectionController.java]: Don't forget to add a validator for rest CreateFullDetection

File: src/main/resources/db/migration/V43_53__Update_ZTJ_and_ZDJ_table.sql [hunk 24]
```diff
@@ -0,0 +1,2 @@
+alter table if exists zone_tiling_job add column if not exists end_to_end_id varchar;
```
Comment (hei-ryan) [inline src/main/resources/db/migration/V43_53__Update_ZTJ_and_ZDJ_table.sql]: End to end ID is used when the ZTJ or ZDJ has to be exposed outside the geo-jobs api. 

Here, it's the full detection ID that must have an end to end it, not the ZTJ or the ZDJ. 

The client doesn't know and SHOULD NOT know, the domain ids, other than the end to end ID.

What you must do then is to just associate the fullDetection entity with the ZTJ and same for ZDJ

File: src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java (hunk 9 above)
Comment (hei-ryan) [inline src/main/java/app/bpartners/geojobs/endpoint/event/model/TilingTaskCreated.java:26]: No. 

The FullDetection must be persisted when the request is handled, and if necessary, get it from the ZTJ attributes associated to the TilingTask, not directly here. Can you explicit why do we need it here ?

File: src/main/java/app/bpartners/geojobs/repository/ZoneTilingJobRepository.java [hunk 25]
```diff
@@ -4,4 +4,6 @@
 import org.springframework.stereotype.Repository;
 
 @Repository
-public interface ZoneTilingJobRepository extends JobRepository<ZoneTilingJob> {}
+public interface ZoneTilingJobRepository extends JobRepository<ZoneTilingJob> {
+  ZoneTilingJob findByEndToEndId(String endToEndId);
```
Comment (hei-ryan) [inline src/main/java/app/bpartners/geojobs/repository/ZoneTilingJobRepository.java]: No.

**Review by hei-ryan — COMMENTED** [review]
Globally, I don't know where do you persist the fullDetection entity ?

**Pull request merged**